tx-parser/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+extern crate edn;
+extern crate combine;
+extern crate mentat_tx;
+
+use combine::{any, eof, many, optional, parser, satisfy_map, token, Parser, ParseResult, Stream};
+use combine::combinator::{Expected, FnParser};
+// TODO: understand why this is self::edn rather than just edn.
+use self::edn::types::Value;
+use mentat_tx::entities::*;
+
+// TODO: implement combine::Positioner on Value.  We can't do this
+// right now because Value is defined in edn and the trait is defined
+// in combine.
+
+// #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+// pub struct ValuePosition {
+//     pub position: usize,
+// }
+
+// impl fmt::Display for ValuePosition {
+//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+//         write!(f, "value position: {}", self.position)
+//     }
+// }
+
+// impl combine::primitives::Positioner for Value {
+//     type Position = ValuePosition;
+//     fn start() -> ValuePosition {
+//         ValuePosition { position: 1 }
+//     }
+//     fn update(&self, position: &mut ValuePosition) {
+//         position.position += 1;
+//     }
+// }
+
+struct Tx<I>(::std::marker::PhantomData<fn(I) -> I>);
+
+type TxParser<O, I> = Expected<FnParser<I, fn(I) -> ParseResult<O, I>>>;
+
+fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser<O, I>
+    where I: Stream<Item = Value>
+{
+    parser(f).expected(err)
+}
+
+impl<I> Tx<I>
+    where I: Stream<Item = Value>
+{
+    fn integer() -> TxParser<i64, I> {
+        fn_parser(Tx::<I>::integer_, "integer")
+    }
+
+    fn integer_(input: I) -> ParseResult<i64, I> {
+        return satisfy_map(|x: Value| if let Value::Integer(y) = x {
```

Comment:
`combinator.rs` is tricky; it took me three days of digging to really understand what is happening :)